docs_src/design/adr/0006-Service-Registry.md

```diff
@@ -0,0 +1,113 @@
+The default service registry used by EdgeX is [Consul](https://github.com/hashicorp/consul) from Hashicorp. Consul is also
+the default configuration provider for EdgeX.
+
+This ADR is meant to address the current usage of the registry by EdgeX services, and in particular whether the EdgeX services
+are using the registry (when enabled) to determine the location of peer services vs. using static per-service configuration.
+The reason this is being investigated is that there's been a proposal that EdgeX do away the registry functionality, as the
+current implementation is not considered "secure" due to current configuration of Consul in an EdgeX deployment as of the
+Hanoi release of EdgeX.
+
+According to the original Service Name Design document (v6)[https://wiki.edgexfoundry.org/display/FA/Architecture+Issues+and+Decisions?preview=%2F7602423%2F12124493%2FServiceNameDesign-v6.pdf], all EdgeX Foundry microservices should be able to accomplish the following tasks:
+
+ * Register with the configuration/registration (referred to simply as “the registry” for the rest of this document) provider (today Consul)
+ * Respond to shutdown requestsoTaking care to clean up of any resources in an orderly fashionoUnregister itself from the registry
+ * _Get the address to another EdgeX microservice by service name through theregistry (when enabled)_
+
+The purpose of this design is to ensure that services themselves advertise their location to the rest of the system by first self-
+registering. Most service registries (including Consul) implement some sort of health check mechanism. If a service is failing one
+or more health checks, the registry will stop reporting its availability when queried.
+
+*Note* - the design specifically excludes device services from this service lookup, as Core Metadata maintains a persistent store of
+DeviceService objects which provide service location.
+
+## Existing Behavior (Hanoi)
+
+### device-virtual-go
+I tested device-virtual-go's behavior by doing the following:
+
+$ sudo snap install edgexfoundry --edge
+$ cp /var/snap/edgexfoundry/current/config/device-virtual/res/configuration.toml .
+
+I edited the file and removed the ```[Client.Data]``` section completely and copied the
+file back into place. Next I enabled device-virtual while monitoring the journal output.
+
+$ sudo cp configuration.toml /var/snap/edgexfoundry/current/config/device-virtual/res/
+$ sudo snap set edgexfoundry device-virtual=on
+
+The following error was seen in the journal:
+
+```
+level=INFO ts=2020-11-13T23:18:17.159313493Z app=device-virtual source=httpserver.go:94 msg="Web server starting (0.0.0.0:49990)"
+error: fatal error; Host setting for Core Data client not configured
+```
+
+Next I followed the same steps, but instead of completely removing the client, I instead set the configuration items in this subsection
+to empty strings or 0 in the case of Port. The same error as above occured, and the service exited.
+
```

Comment:
Is this functionality baked into the SDK or just the virtual service?  Is the current state of the C device services and SDK the same?

Reply:
This functionality is implemented in device-sdk-go (see [internal/clients/init.go](https://github.com/edgexfoundry/device-sdk-go/blob/93569aefbcf2dcd4d18a31fcf2c190494ae30d88/internal/clients/init.go#L54)).

As for device-sdk-c, I took a cursory look, and it appears it actually does query the registry for the host and port of core data and metadata (see [src/c/service.c](https://github.com/edgexfoundry/device-sdk-c/blob/master/src/c/service.c#L730)) if the registry flag has been set, otherwise it uses the clients from `configuration.toml`. @iain-anderson can you please confirm?